// most-subject/lib/commonjs/sources/HoldSubjectSource.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var most_1 = require("most");
var multicast_1 = require("@most/multicast");
var prelude_1 = require("@most/prelude");
var HoldSubjectSource = (function (_super) {
    __extends(HoldSubjectSource, _super);
    function HoldSubjectSource(source, bufferSize) {
        var _this = _super.call(this, source) || this;
        _this.has = false;
        _this.buffer = [];
        _this.bufferSize = bufferSize;
        return _this;
    }
    HoldSubjectSource.prototype.add = function (sink) {
        if (this.has) {
            pushEvents(this.buffer, sink);
        }
        return _super.prototype.add.call(this, sink);
    };
    HoldSubjectSource.prototype.event = function (time, value) {
        this.has = true;
        this.buffer = dropAndAppend(value, this.buffer, this.bufferSize);
        return _super.prototype.event.call(this, time, value);
    };
    return HoldSubjectSource;
}(multicast_1.MulticastSource));
exports.HoldSubjectSource = HoldSubjectSource;
function pushEvents(buffer, sink) {
    var length = buffer.length;
    for (var i = 0; i < length; ++i) {
        sink.event(most_1.defaultScheduler.now(), buffer[i]);
    }
}
function dropAndAppend(value, buffer, bufferSize) {
    if (buffer.length === bufferSize) {
        return prelude_1.append(value, prelude_1.drop(1, buffer));
    }
    return prelude_1.append(value, buffer);
}
exports.dropAndAppend = dropAndAppend;
