// kafkajs-stream/module/ConsumerStream.js
import { Readable } from "stream";
export class ConsumerStream extends Readable {
    constructor(kafka, options) {
        super();
        this.onCrash = async (err) => {
            console.error(err);
            this.init();
            await this.start();
        };
        this.kafka = kafka;
        this.config = options.config;
        this.runConfig = options.runConfig;
        this.topic = options.topic;
        this.init();
    }
    init() {
        this.connected = false;
        this.started = false;
        this.paused = false;
    }
    _read() {
        (async () => {
            try {
                await this.start();
            }
            catch (e) {
                this.destroy(e);
            }
        })();
    }
    async start() {
        if (!this.connected) {
            this.connected = true;
            this.consumer = this.kafka.consumer(this.config);
            await this.consumer.connect();
            await this.consumer.subscribe(this.topic);
            this.consumer.on("consumer.crash", this.onCrash);
        }
        if (!this.started) {
            this.started = true;
            await this.run();
        }
        if (this.paused) {
            this.paused = false;
        }
    }
    async run() {
        await this.consumer.run({
            eachBatchAutoResolve: false,
            eachBatch: async ({ batch, resolveOffset, heartbeat }) => {
                if (this.paused) {
                    return;
                }
                for (const message of batch.messages) {
                    if (this.paused) {
                        break;
                    }
                    const continueToPush = this.push(message.value);
                    resolveOffset(message.offset);
                    await heartbeat();
                    if (!continueToPush) {
                        this.paused = true;
                    }
                }
            },
            ...this.runConfig,
        });
    }
    _destroy(error) {
        this.consumer.disconnect();
        super.destroy(error === null ? undefined : error);
    }
}
